File: mirror/tier2-ruby/showcase/perf_benchmark.rb
# MIRROR V3: Tier 2 Ruby Showcase
# Category: Optimization
# Module: perf_benchmark.rb
# Purpose: Demonstrate performance benchmarking

def measure_time(method_sym = nil, iterations: 1000, &block)
  start = Time.now
  
  iterations.times do
    block.call if block
  end
  
  elapsed = (Time.now - start) * 1000
  
  {
    elapsed_ms: elapsed.round(3),
    iterations: iterations,
    per_call_us: (elapsed * 1000 / iterations).round(2),
  }
end

def simple_operation(x)
  x * 2
end

def complex_operation(x)
  result = 0
  10.times { |i| result += x * i }
  result
end

def optimized_operation(x)
  x * 10 * 9 / 2
end

simple_bench = measure_time(iterations: 10000) { simple_operation(100) }
complex_bench = measure_time(iterations: 1000) { complex_operation(100) }
optimized_bench = measure_time(iterations: 10000) { optimized_operation(100) }

result = {
  simple_elapsed: simple_bench[:elapsed_ms],
  complex_elapsed: complex_bench[:elapsed_ms],
  optimized_elapsed: optimized_bench[:elapsed_ms],
  benchmarking_active: true,
}

puts result.inspect

File: mirror/tier2-dart/showcase/perf_benchmark.dart
// MIRROR V3: Tier 2 Dart Showcase
// Category: Optimization
// Module: perf_benchmark.dart

Map<String, dynamic> measureTime(Function func, {int iterations = 1000}) {
  final start = DateTime.now().millisecondsSinceEpoch;
  
  for (int i = 0; i < iterations; i++) {
    func();
  }
  
  final elapsed = (DateTime.now().millisecondsSinceEpoch - start).toDouble();
  
  return {
    'elapsedMs': elapsed,
    'iterations': iterations,
    'perCallUs': (elapsed * 1000 / iterations),
  };
}

int simpleOperation(int x) => x * 2;

int complexOperation(int x) {
  int result = 0;
  for (int i = 0; i < 10; i++) {
    result += x * i;
  }
  return result;
}

int optimizedOperation(int x) => (x * 10 * 9 ~/ 2);

void main() {
  final simpleBench = measureTime(() => simpleOperation(100), iterations: 10000);
  final complexBench = measureTime(() => complexOperation(100), iterations: 1000);
  final optimizedBench = measureTime(() => optimizedOperation(100), iterations: 10000);
  
  final result = <String, dynamic>{
    'simpleElapsed': simpleBench['elapsedMs'],
    'complexElapsed': complexBench['elapsedMs'],
    'optimizedElapsed': optimizedBench['elapsedMs'],
    'benchmarkingActive': true,
  };
  print(result);
}
